system/security/shared/caching.cpp

```diff
@@ -697,35 +721,38 @@ SecAccessFlags CPermissionsCache::queryDefaultPermission(ISecUser& user)
 
     SecAccessFlags defaultPermission = SecAccess_None;
     const std::string username(user.getName());
-    bool addedToCache = false;
+
     {
-        CriticalBlock defaultScopePermissionBlock(syncDefaultScopePermissions);
+        ReadLockBlock readLock(m_defaultFilePermissionRWLock);
         auto it = m_userDefaultFileScopePermissions.find(username);
+        readLock.clear();   // lock no longer needed
+
+        // If default permission not found for the user, query the security manager.
+        // Then, acquire the write lock and add the default permission to the cache if
+        // not already added by another thread.
         if (it == m_userDefaultFileScopePermissions.end())
         {
             defaultPermission = m_secMgr->queryDefaultPermission(user);
-            m_userDefaultFileScopePermissions.emplace(username, defaultPermission);
-            addedToCache = true;
+            WriteLockBlock writeLock(m_defaultFilePermissionRWLock);
+            it = m_userDefaultFileScopePermissions.find(username);
+            if (it == m_userDefaultFileScopePermissions.end())
+            {
+                m_userDefaultFileScopePermissions.emplace(username, defaultPermission);
+                DBGLOG("Added user '%s' to default file scope permissions with access %s(%d)", username.c_str(), getSecAccessFlagName(defaultPermission),
```

Comment:
Previous code was careful to keep this logging outside the critsec - you should do the same here for the lock